### PR TITLE
validator: Add `--enable-bigtable-ledger-upload` flag 

### PR DESCRIPTION
**Pull request by mvines**
When enabled, the `--enable-bigtable-ledger-upload` flag will cause the validator to continually upload new confirmed blocks to BigTable in the background.

### CODE REVIEW

File: core/src/rpc_service.rs [hunk 1]
```diff
@@ -260,20 +263,37 @@ impl JsonRpcService {
...
-                .unwrap_or_else(|err| {
-                    error!("Failed to initialize BigTable ledger storage: {:?}", err);
-                    None
-                })
-        } else {
-            None
-        };
+        let exit_bigtable_ledger_upload_service = Arc::new(AtomicBool::new(false));
+
+        let (bigtable_ledger_storage, _bigtable_ledger_upload_service) =
```
Comment (mvines) [inline core/src/rpc_service.rs:268]: It's slightly awkward that `BigTableUploadService` is plumbed into `RpcService` like this.   Moving it up a level into `Validator` would require moving the tokio Runtime and `solana_storage_bigtable::LedgerStorage::new` as well and I didn't like that code churn in this PR.    That's my excuse at least

File: ledger-tool/src/bigtable.rs [hunk 2]
```diff
@@ -1,218 +1,39 @@
...
-// Read up to this many blocks from blockstore before blocking on the upload process
-const BLOCK_READ_AHEAD_DEPTH: usize = NUM_BLOCKS_TO_UPLOAD_IN_PARALLEL * 2;
+use std::{
+    path::Path,
+    process::exit,
+    result::Result,
+    sync::{atomic::AtomicBool, Arc},
+};
 
 async fn upload(
```
Comment (mvines) [inline ledger-tool/src/bigtable.rs:18]: Changes in this file are really just moving code into `ledger/`

**Comment by mvines** [comment]
For mainnet-beta, I think we keep the same upload at the end of an epoch process going as the last resort.  Then additionally add this flag on one or more of the warehouse nodes to get confirmed block data into BigTable within seconds.   At that point, we can start making API nodes ephemeral!

**Review by CriesofCarrots — COMMENTED** [review]
Looks good overall. I'll buy your rationalization for the new service in rpc_service, though I agree it's awkward.

I'm wondering about 2 potential races:

1. Is this the right code for the end-of-epoch upload? https://github.com/solana-labs/cluster/blob/3614bbf6d36a275f4a5c0cba10d517367825b508/bin/warehouse-upload-to-storage-bucket.sh#L74 It seems like that is checking the entire ledger for gaps in bigtable storage all at once, and won't catch new confirmed blocks that are uploaded by the BigTableUploadService in the meantime. ... on looking at the lower-level apis more, maybe this isn't an issue, beyond efficiency.

2. Although this isn't merged yet, it will bring the complication of blocktimes and might be worth planning for now: https://github.com/solana-labs/solana/pull/11955. Because the CacheBlockTimeService is prompted by `handle_votable_bank` and the BigTableUploadService by a changes in `highest_confirmed_root`, if both PRs get merged as is, the blocktime "should" get cached before the upload happens. But if the blocktime calculation backs up for some reason, there's no guarantee.

**Comment by mvines** [comment]
> 1. Is this the right code for the end-of-epoch upload? https://github.com/solana-labs/cluster/blob/3614bbf6d36a275f4a5c0cba10d517367825b508/bin/warehouse-upload-to-storage-bucket.sh#L74 It seems like that is checking the entire ledger for gaps in bigtable storage all at once, and won't catch new confirmed blocks that are uploaded by the BigTableUploadService in the meantime. ... on looking at the lower-level apis more, maybe this isn't an issue, beyond efficiency.

Yes that's the right code.   But this always runs on a ledger from the previous Epoch, using the old ledger from the warehouse node.  In the meantime, the warehouse node has been restarted from a snapshot and is building a new ledger.

> 2. Although this isn't merged yet, it will bring the complication of blocktimes and might be worth planning for now: #11955. Because the CacheBlockTimeService is prompted by `handle_votable_bank` and the BigTableUploadService by a changes in `highest_confirmed_root`, if both PRs get merged as is, the blocktime "should" get cached before the upload happens. But if the blocktime calculation backs up for some reason, there's no guarantee.

I don't think we need to be this aggressive with uploading ledger into BigTable.  Would it help to trail the highest confirmed root by 100 slots or so?

File: ledger/src/bigtable_upload.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    };
+
+    // The blocks that still need to be uploaded is the difference between what's already in the
+    // bigtable and what's in blockstore...
+    let blocks_to_upload = {
+        let blockstore_slots = blockstore_slots.iter().cloned().collect::<HashSet<_>>();
+        let bigtable_slots = bigtable_slots.into_iter().collect::<HashSet<_>>();
+
+        let mut blocks_to_upload = blockstore_slots
+            .difference(&blockstore_slots_with_no_confirmed_block)
```
[the remark this answers is not in the document]
Comment (mvines) [reply ledger/src/bigtable_upload.rs:109]: yeah that looks like dead code from an earlier iteration.  I'll get rid of it thanks

**Comment by CriesofCarrots** [comment]
> Yes that's the right code. But this always runs on a ledger from the previous Epoch, using the old ledger from the warehouse node. In the meantime, the warehouse node has been restarted from a snapshot and is building a new ledger.

Aha! That explains it. Thanks for that detail.

> I don't think we need to be this aggressive with uploading ledger into BigTable. Would it help to trail the highest confirmed root by 100 slots or so?

Yep, that would definitely help. 100 slots should be more than enough

**Comment by CriesofCarrots** [comment]
> Maybe once #11955 lands, the loop in here can check if `blockstore.get_block_time()` returns Some(_) for the highest slot it wants to upload. If None not then wait a little while longer

Sounds good. I’ll play with some options when I rebase #11955 on this.

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**